### ISSUE TITLE
Interoperability with cats

### ISSUE THREAD

**Issue by ShahOdin** [umazalakain/errata#6]
Hey! Love the idea and it makes me wonder why the distinction between `Raise` and `HandleTo` wasn't done already! :) 

Not sure what the end game for this repo is and if it's more of a proof of concept or if it aims to be something more, but thought I'd take this for a test run and apply it to an implementation in a cats-based code base.

# Refactor eixisting cats-based code: case study 

## Missing cats Syntax

I was looking at a simple enough piece of code which essentially translated to this:

```scala
def apply[F[_]: ApplicativeThrow] = Right[Throwable, String]("foo").liftTo[F]
``` 

with [liftTo](https://github.com/typelevel/cats/blob/31084ac70d5610bb83fe41b9a2bdc09905f05c63/core/src/main/scala/cats/syntax/either.scala#L355) relying on `ApplicativeError`'s [fromEither](https://github.com/typelevel/cats/blob/be8a9930a439aec2ae31f8c1f044843ab43e0be8/core/src/main/scala/cats/ApplicativeError.scala#L314). 

I can see we already have the following: https://github.com/umazalakain/errata/blob/4e3199baaedc6734a6e9c48883ce4fa544dee979/errata/src/main/scala/errata/types.scala#L33

be nice to have the `liftTo` part as well. There are a few other ones such as:
- https://github.com/typelevel/cats/blob/31084ac70d5610bb83fe41b9a2bdc09905f05c63/core/src/main/scala/cats/syntax/option.scala#L356
- https://github.com/typelevel/cats/blob/31084ac70d5610bb83fe41b9a2bdc09905f05c63/core/src/main/scala/cats/syntax/option.scala#L376

## Missing cats instances

There is also a bunch of cats instances that depend on an `ApplicativeErorr` and if I were to switch to using `Raise` / `HandleTo`, as it stands I would no longer be able to use the flexibility that cats provide me. Perhaps the equivalent of those should also be added? 

Think adding the above would allow folks to use this lib more easily together with cats. 

# Similar ideas explored elsewhere?

While searching for the `ApplicativeErorr` dependent helper methods in my project's scope I came across: 

- cats-mtl having similar concepts of: https://typelevel.org/cats-mtl/mtl-classes/handle.html and https://typelevel.org/cats-mtl/mtl-classes/raise.html with the slight differnece that their `Handle` extends `Raise`.

- fs2 also has [RaiseThrowable](https://github.com/typelevel/fs2/blob/main/core/shared/src/main/scala/fs2/RaiseThrowable.scala#L36-L44) which seems to be used as a type witness only and is used like [this](https://github.com/typelevel/fs2/blob/cc913dd0f7314257856f7e28fd86017785db8ecc/core/shared/src/main/scala/fs2/Pull.scala#L379) for example (which doesn't seem to actually use `raise`? 😕 ) When it comes to fs2, I'm definitely one of those 80% that only understand 20% of it, if not 2% but unless I'm misunderstanding the role of `RaiseThrowable` I'd imagine fs2's code readability would benefit from using a similar universal concept of `Raise`.

Just wondering how all these would/could tie together.

**Comment by umazalakain** [comment]
Hey Shah!

Thanks for your comments and suggestions, I greatly appreciate them.

# Future work and related projects

Regarding my intention with this project: for one, I would like to start using it at my daytime job, I believe it would be beneficial, but long term, I would aspire to either publish it under the typelevel umbrella, or even integrate it into cats.

As you point out, cats-mtl defines similar types, with two big differences:
- `Handle` extends `Raise` --- you cannot limit yourself to solely handling errors.
- `Handle` is parametrised by a single effect --- you cannot bind an output effect where errors have been handled, a la https://github.com/umazalakain/errata/blob/main/errata/src/main/scala/errata/types.scala#L52.

You are not the first person that asks me about the definitions in cats-mtl and how errata relates to those, do you think it's worth mentioning it in the readme?

I'm not entirely certain about why they use `RaiseThrowable` as a witness in fs2. They construct a `RaiseThrowable[F]` for every `ApplicativeThrow[F]`, so why not use this latter one directly, I guess that in addition to those they also want to be able to build `RaiseThrowable[F]` for some instances where there is no `ApplicativeThrow[F]`? But anyway, it looks like it's merely an engineering artefact, rather than a solid concept in itself.

# Syntax and instances

The `liftTo[F]` syntax is a very good idea, I will add those today and make a new release. 

Regarding cats instances that depend on `ApplicativeError`: I believe in principle https://github.com/umazalakain/errata/blob/main/errata/src/main/scala/errata/instances.scala#L93 makes it unnecessary for us to derive each of them separately. If this is imported with `import errata.instances.inverse.*`, `ApplicativeError[F, E]` should automatically be derived from `Applicative[F]` plus `Errors[F, E]`. Now, I would have to have a look at those instances you mention, because it might be the case that they could be derived without going as much up the tree as `Errors[F, E]` (e.g maybe some only need of `Raise[F, E]`). I'll have a look!

# Refactoring existing cats-based code

I would be _really_ interested in knowing more about how the refactoring goes once the above are solved, please keep me in the loop!

**Comment by ShahOdin** [comment]
> do you think it's worth mentioning it in the readme?

yeah I'd like that.

> Now, I would have to have a look at those instances you mention, because it might be the case that they could be derived without going as much up the tree as Errors[F, E] 

yeah that's exactly what I was gonna say. I'd imagine most of those instances would only need a `Raise` instance.

> please keep me in the loop!

will do ;)

**Comment by umazalakain** [comment]
@ShahOdin I added the syntax and some instances, lmk what you think!

**Comment by ShahOdin** [comment]
Looks good. :)

**Comment by umazalakain** [comment]
Added some comments on cats-mtl and published a new version. Let me know if you find any other impediments!

**Comment by ShahOdin** [comment]
> I would be really interested in knowing more about how the refactoring goes

Tried the new release. some observations:

- I applied the refactor for a single component `Qux` that was a few levels deep inside a dependency tree: `Foo -> Bar -> Baz -> Qux`. 
- Previously `ApplicativeThrow[F]` was passed down from `Foo` all the way down to `Qux`. In my refactor, this has largely remained unchanged and the only changes have been in `Baz` and `Qux`. `import errata.instances.*` in `Baz` came to the rescue. ✅ 
- `Qux` previosuly expected a `MonadThrow` instance. it now expects `Monad: RaiseThrow` with a locally defined: `type RaiseThrow[F[_]] = Raise[F, Throwable]` as a quick fix. `Qux` technically only needs `Raise[F, FooError]` with `case class FooError(...) extends Throwable`. I understand that making `RaiseThrow` available might be considered encouraging an antipattern but I'd argue having it there for a minimal refactor might not be a bad idea.
- re the new syntax, `liftTo[F]` didn't compile and I had to drop `[F]`. I can see why you changed it. I should note that I had to change my `import cats.syntax.all.*` to lots of individual separate ones to avoid extension method ambiguity. not ideal. Guess it's a similar story as `raiseError` vs `raise`? best rename it as a quick fix.

I guess the other approach, if we don't want to come up with all these silly alternative names, is using the namespacing approach that fs2 does with `.compile` in exposing certain API only when you explicitly ask for it. so we could have `myError.withErrata.raiseError` or `myEither.withErrata.liftTo` or similar. maybe an overkill though 🤷🏽 

tldr:

- `RaiseThrow` and `HandleThrow` would be nice to have
- `liftTo` in its current form is tedious to use

**Comment by umazalakain** [comment]
I added the type aliases for `Throwable`. Regarding syntax and namespacing, what do you think about giving the user the option to do either with something like this https://github.com/umazalakain/errata/pull/7

**Comment by umazalakain** [comment]
Regarding `liftTo[F]` vs `liftTo`: I guess the difference is whether we keep the constraint on e.g. `Raise[F, E]` on the implicit class, or if we move it to the method. If we keep it in the implicit class then `F` must be bound there too, so we get `liftTo`. If we move it to the method we can bind `F` locally. The former is more satisfying in a way, since you only get the syntax if you have the goods to justify it. However the latter makes discoverability in IDEs easier, and from what I've seen is the method preferred by cats. Wdyt?

**Comment by ShahOdin** [comment]
> #7

Me likey! 🥇

**Comment by ShahOdin** [comment]
> the latter makes discoverability in IDEs easier

I mean I have never been a library author but specially with the ambitions for this project, and people being used to the IDE-friendly cats, I'd say prioritising the user experience should be key.

**Comment by umazalakain** [comment]
Agreed, I'll do some refactoring

**Comment by umazalakain** [comment]
@ShahOdin https://github.com/umazalakain/errata/pull/8 it looks good, but there's a caveat: the user must now supply type parameters. I'm a little conflicted

**Comment by umazalakain** [comment]
Closing this one and opening a new ticket for simultanous use of syntax with and without namespace
